src/provider/pedersen.rs

```diff
@@ -182,115 +178,97 @@ impl<G: Group> CommitmentEngineTrait<G> for CommitmentEngine<G> {
   }
 
   fn commit(ck: &Self::CommitmentKey, v: &[G::Scalar]) -> Self::Commitment {
+    if ck.ck.len() < v.len() {
+      println!("commitment key length: {}", ck.ck.len());
+      println!("v length: {}", v.len());
+    }
     assert!(ck.ck.len() >= v.len());
     Commitment {
       comm: G::vartime_multiscalar_mul(v, &ck.ck[..v.len()]),
     }
   }
 }
 
-/// A trait listing properties of a commitment key that can be managed in a divide-and-conquer fashion
-pub trait CommitmentKeyExtTrait<G: Group> {
-  /// Holds the type of the commitment engine
-  type CE: CommitmentEngineTrait<G>;
-
+/// Additional extensions on the commitment engine
+pub trait CommitmentEngineExtTrait<G: Group>: CommitmentEngineTrait<G> {
   /// Splits the commitment key into two pieces at a specified point
-  fn split_at(&self, n: usize) -> (Self, Self)
-  where
-    Self: Sized;
+  fn split_at(ck: &Self::CommitmentKey, n: usize) -> (Self::CommitmentKey, Self::CommitmentKey);
 
   /// Combines two commitment keys into one
-  fn combine(&self, other: &Self) -> Self;
+  fn combine(ck: &Self::CommitmentKey, other: &Self::CommitmentKey) -> Self::CommitmentKey;
 
   /// Folds the two commitment keys into one using the provided weights
-  fn fold(&self, w1: &G::Scalar, w2: &G::Scalar) -> Self;
+  fn fold(ck: &Self::CommitmentKey, w1: &G::Scalar, w2: &G::Scalar) -> Self::CommitmentKey;
 
   /// Scales the commitment key using the provided scalar
-  fn scale(&self, r: &G::Scalar) -> Self;
-
-  /// Reinterprets commitments as commitment keys
-  fn reinterpret_commitments_as_ck(
-    c: &[<<<Self as CommitmentKeyExtTrait<G>>::CE as CommitmentEngineTrait<G>>::Commitment as CommitmentTrait<G>>::CompressedCommitment],
-  ) -> Result<Self, SpartanError>
-  where
-    Self: Sized;
-}
+  fn scale(ck: &Self::CommitmentKey, r: &G::Scalar) -> Self::CommitmentKey;
 
-impl<G: Group> CommitmentKeyExtTrait<G> for CommitmentKey<G> {
-  type CE = CommitmentEngine<G>;
+  /// Reinterprets the commitments as a commitment key
```

Comment:
It would be really good to have a line of comments or two explaining the re-interpretation and a justification for why it makes sense (naturally, this wasn't introduced in your PR, but relating the `CommitmentKeyExtTrait` to homomorphic properties is non-trivial enough to deserve a comment IMHO).